File: src/com/ptby/dynamicreturntypeplugin/index/ClassConstantAnalyzer.java
package com.ptby.dynamicreturntypeplugin.index;

import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.PsiElementBase;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.psi.elements.Field;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.resolve.types.PhpType;

import java.util.Collection;

public class ClassConstantAnalyzer {


    private String castClassConstantToPhpType( PsiElementBase elementBase, PsiElement element, String classConstant ) {
        String[] constantParts = classConstant.split( "(#K#C|\\.|\\|\\?)" );
        if ( constantParts.length < 3 ) {
            return null;
        }

        String className = constantParts[ 1 ];
        String constantName = constantParts[ 2 ];

        PhpIndex phpIndex = PhpIndex.getInstance( elementBase.getProject() );
        Collection<PhpClass> classesByFQN = phpIndex.getClassesByFQN( className );
        for ( PhpClass phpClass : classesByFQN ) {
            Collection<Field> fields = phpClass.getFields();
            for ( Field field : fields ) {
                if ( field.isConstant() && field.getName().equals( constantName ) ) {
                    PsiElement defaultValue = field.getDefaultValue();
                    if ( defaultValue == null ) {
                        return null;
                    }
                    String constantText = defaultValue.getText();
                    if ( constantText.equals( "__CLASS__" ) ) {
                        PhpType phpType = new PhpType();
                        phpType.add( className );
                        return className;
                    }
                }
            }
        }

        return null;
    }

}

File: src/com/ptby/dynamicreturntypeplugin/index/ClassConstantAnalyzer.kt
package com.ptby.dynamicreturntypeplugin.index

import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.jetbrains.php.PhpIndex
import com.jetbrains.php.lang.psi.elements.Field
import com.jetbrains.php.lang.psi.elements.PhpClass
import com.ptby.dynamicreturntypeplugin.signature_extension.*
import com.ptby.dynamicreturntypeplugin.signatureconversion.SignatureMatcher


/**
 * @deprecated - use ClassConstantWalker directly
 */
public class ClassConstantAnalyzer {

    public fun getClassNameFromConstantLookup(classConstant: String, project: Project): String? {
        val classConstantWalker  = ClassConstantWalker()
        val result = classConstantWalker.walkThroughConstants(PhpIndex.getInstance(project), classConstant)
        return result
    }
}
